docs: Add Prisma documentation (#1404)
TypedProject/ts-express-decorators: docs/tutorials/prisma.md

It is used as an alternative to writing plain SQL, or using another database access tool such as SQL query builders (like [knex.js](/tutorials/objection.md)) or ORMs (like [TypeORM](/tutorials/typeorm.md) and [Sequelize](https://sequelize.org/)). 
Prisma currently supports PostgreSQL, MySQL, SQL Server and SQLite.

While Prisma can be used with plain JavaScript, it embraces TypeScript and provides a level to type-safety that goes beyond the guarantees other ORMs in the TypeScript ecosystem.
You can find an in-depth comparison of the type-safety guarantees of Prisma and TypeORM [here](https://www.prisma.io/docs/concepts/more/comparisons/prisma-and-typeorm#type-safety).

See issue for more details: https://github.com/tsedio/tsed/issues/1389

::: tip Note
If you want to get a quick overview of how Prisma works, you can follow the [Quickstart](https://www.prisma.io/docs/getting-started/quickstart) or read the [Introduction](https://www.prisma.io/docs/understand-prisma/introduction) in the [documentation](https://www.prisma.io/docs/).
:::

<Projects type="projects"/>

## Prisma generator for Ts.ED <Badge text="Premium sponsors" />

> If you are premium sponsors, or you want to become it, you can ask [Ts.ED team](https://api.tsed.io/rest/slack/tsedio/tsed) to get your access to the private packages [`@tsedio/prisma`](/tutorials/prisma-client.md). 

Reviewer: ```suggestion
If you are a premium sponsor, or you want to become one, you can ask [Ts.ED team](https://api.tsed.io/rest/slack/tsedio/tsed) to get your access to the private packages [`@tsedio/prisma`](/tutorials/prisma-client.md). 
```

> This package generates enums and classes compatible with the Ts.ED decorators like @@Returns@@ and extend the possibility about the `prisma.schema`.

Reviewer: ```suggestion
This package generates enums and classes compatible with the Ts.ED decorators like @@Returns@@ and extends possibilities about the `prisma.schema`.
```

> The following tutorial is to be followed to create your project. You will have more details regarding the installation of the [`@tsedio/prisma`](/tutorials/prisma-client.md) package.

Reviewer: ```suggestion
You should follow this tutorial to create your project properly. You will have more details regarding the installation of the [`@tsedio/prisma`](/tutorials/prisma-client.md) package.
```

For the purpose of this guide, you'll use a [SQLite](https://sqlite.org/) database to save the overhead of setting up a database server.

Note that you can still follow this guide, even if you're using PostgreSQL or MySQL – you'll get extra instructions for using these databases at the right places.

::: tip Note
If you already have an existing project and consider migrating to Prisma, you can follow the guide for [adding Prisma to an existing project](https://www.prisma.io/docs/getting-started/setup-prisma/add-to-existing-project-typescript-postgres). If you are migrating from TypeORM, you can read the guide [Migrating from TypeORM to Prisma](https://www.prisma.io/docs/guides/migrate-to-prisma/migrate-from-typeorm). 
:::

## Create a Ts.ED project

To get started, install the Ts.ED CLI and create your app skeleton with the following commands:

```sh
npm install -g @tsed/cli
tsed init tsed-prisma
```
From the CLI, select the following features: Swagger, Eslint, Jest.

## Set up Prisma

Start by installing the Prisma CLI as a development dependency in your project:
```sh
cd tsed-prisma

> npm install --save-dev prima

Reviewer: t'es sûr que c'est "prima" et pas "prisma" ?

Author: prisma

```

> In the following steps, we'll be utilizing the [Prisma CLI](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-cli). 

Reviewer: we'll be using ?

Author: Ok :)

Prisma Migrate generates SQL migration files for your declarative data model definition in the Prisma schema. 
These migration files are fully customizable so that you can configure any additional features of the underlying database or include additional commands, e.g. for seeding.

Add the following two models to your `schema.prisma` file:

```groovy
model User {
  id    Int     @default(autoincrement()) @id
  email String  @unique
  name  String?
  posts Post[]
}

model Post {
  id        Int      @default(autoincrement()) @id
  title     String
  content   String?
  published Boolean? @default(false)
  author    User?    @relation(fields: [authorId], references: [id])
  authorId  Int?
}
```

With your Prisma models in place, you can generate your SQL migration files and run them against the database. 
Run the following commands in your terminal:

```sh
npx prisma migrate dev --name init
```

::: tip
Add the previous command to your `scripts` in the `package.json`:

```json
{
  "scripts": {
    "prisma:migrate": "npx prisma migrate dev --name init"
  }
}
```

:::

This `prisma migrate dev` command generates SQL files and directly runs them against the database. 

> In this case, the following migration files was created in the existing `prisma` directory:

Reviewer: ```suggestion
In this case, the following migration file was created in the existing `prisma` directory:
```

```bash
$ tree prisma
prisma
├── dev.db
├── migrations
│   └── 20210507100915_init
│       └── migration.sql
└── schema.prisma
```

### Install and generate Prisma Client

Prisma Client is a type-safe database client that's _generated_ from your Prisma model definition.
Because of this approach, Prisma Client can expose [CRUD](https://www.prisma.io/docs/concepts/components/prisma-client/crud) operations that are _tailored_ specifically to your models.

To install Prisma Client in your project, run the following command in your terminal:

```bash
$ npm install @prisma/client
```

Note that during installation, Prisma automatically invokes the `prisma generate` command for you.
In the future, you need to run this command after _every_ change to your Prisma models to update your generated Prisma Client.

The `prisma generate` command reads your Prisma schema and updates the generated Prisma Client library inside `node_modules/@prisma/client`.

::: warning

> The generated interfaces cannot be used with `@Returns` to describe the Swagger documentation! You have to use the premium packages [`@tsedio/prisma`](/tutorials/prisma-client.md)

Reviewer: ```suggestion
The generated interfaces cannot be used with `@Returns` to describe the Swagger documentation! You have to use the premium package [`@tsedio/prisma`](/tutorials/prisma-client.md)
```

:::

::: tip
Add the previous command to your `scripts` in the `package.json`:

```json
{
  "scripts": {
    "prisma:generate": "npx prisma generate"
  }
}
```
:::

## Configure Ts.ED prisma client <Badge text="Premium sponsors" />

> Now our `prisma.schema` is ready for the basic usage. It's time to configure the [`@tsedio/prisma`](/tutorials/prisma-client.md)

Reviewer: ```suggestion
Now our `prisma.schema` is ready for the basic usage. It's time to configure [`@tsedio/prisma`](/tutorials/prisma-client.md)
```

  }): Promise<Post[]> {
    const {skip, take, cursor, where, orderBy} = params;
    return this.prisma.post.findMany({
      skip,
      take,
      cursor,
      where,
      orderBy
    });
  }

  async createPost(data: Prisma.PostCreateInput): Promise<Post> {
    return this.prisma.post.create({
      data
    });
  }

  async updatePost(params: {where: Prisma.PostWhereUniqueInput; data: Prisma.PostUpdateInput}): Promise<Post> {
    const {data, where} = params;
    return this.prisma.post.update({
      data,
      where
    });
  }

  async deletePost(where: Prisma.PostWhereUniqueInput): Promise<Post> {
    return this.prisma.post.delete({
      where
    });
  }
}
```

Your `UserService` and `PostService` currently wrap the CRUD queries that are available in Prisma Client. 
In a real world application, the service would also be the place to add business logic to your application. 
For example, you could have a method called `updatePassword` inside the `UserService` that would be responsible for updating the password of a user.

## Create controllers

Finally, you'll use the services you created in the previous sections to implement the different routes of your app. 

> Now we have to create controllers to exposes your business to our consumers. So create the following controllers in `src/controllers` directory:

Reviewer: ```suggestion
Now we have to create controllers to expose your business to our consumers. So create the following controllers in `src/controllers` directory:
```

    if (!model) {
      throw new NotFound("Post not found");
    }

    return model;
  }

  @Get("/search/:searchString")
  @(Returns(200, Array).Of(PostModel))
  async getFilteredPosts(@PathParams("searchString") searchString: string): Promise<PostModel[]> {
    return this.service.posts({
      where: {
        OR: [
          {
            title: {contains: searchString}
          },
          {
            content: {contains: searchString}
          }
        ]
      }
    });
  }
}
```
  </Tab>
  <Tab label="">

```typescript
import {Controller, Get} from "@tsed/common";
import {Inject} from "@tsed/di";
import {PostModel} from "@tsedio/prisma"; // Remove it if you haven't the @tsedio/prisma package access!
import {Returns} from "@tsed/schema";
import {PostService} from "../services/PostService";

@Controller("/feeds")
export class FeedsController {
  @Inject()
  service: PostService;

  @Get("/")
  @(Returns(200, Array).Of(PostModel))
  getFeeds() {
    return this.service.posts({
      where: {published: true}
    });
  }
}
```

  </Tab>
</Tabs>

### Summary

In this tutorial, you learned how to use Prisma along with Ts.ED to implement a REST API. 

> The controller that implements the routes of the API is calling a `PrismaService` which in turn uses Prisma Client to send queries to a database to fulfill the data needs of incoming requests.

Reviewer: ```suggestion
The controller implementing the routes of the API is calling a `PrismaService` which in turn uses Prisma Client to send queries to a database to fulfill the data needed by incoming requests.
```
le "needed by" je suis pas sûre que c'était ce que tu voulais dire, mais la phrase d'origine n'a pas de sens :P

Author: C'est copié d'un autre tuto...
